### DISCUSSION TITLE
A premixed flame case with EB. "deltaT_iters not converged"

### DISCUSSION THREAD

**Discussion by ZhaofanZhu** [AMReX-Combustion/PeleLMeX#242] [Q&A]
Hi, I am trying a case in which the premixed flame propagating along a rough wall. The rough wall is a 'EB2::PlaneIF' with sinusoidal shape.

Firstly I set the amplitude of the sine wave to 0. The wall becomes 'smooth' and the case worked as expected:

![image](https://github.com/AMReX-Combustion/PeleLMeX/assets/144213811/7d08bf46-bb7e-48e5-87a9-0ff438eb3ef8)

Then I change the amplitude of the sine wave to non-zero:

![image](https://github.com/AMReX-Combustion/PeleLMeX/assets/144213811/224306ac-85e6-4f23-8378-b0fbd7fee74d)

The case not converged, even though I set a high 'peleLM.deltaT_tol = 1e-4'. Moreover, the divergence happens very early, when the initialized high-temperature region has not yet been affected by the wall, which makes me very puzzled.

![image](https://github.com/AMReX-Combustion/PeleLMeX/assets/144213811/b113a0e6-33d4-4df8-93ab-c758794d23f9)

I am looking forward to any reply. Thanks in advance!

**Comment by baperry2** [comment]
This is indeed puzzling. Because it works fine in the otherwise identical case with the flat wall, it doesn't seem related to the thermo properties issue discussed in #238 . We don't usually run cases with as small EB features as this; my expectation is that the multigrid linear solvers used for the pressure projection in the low Mach number algorithm and for the implicit diffusion could have a bit more trouble with such small features, but that doesn't seem like what's happening here.

Here are a few random ideas of thing you could try to investigate the issue:
- run with `diffusion.verbose = 2`, which will print diagnostic output for the linear solves for diffusion to verify that these are converging properly
- use the `WriteDebugPlotFile` function to output plots of intermediate quantities during the deltaT iterations, and try to figure out exactly where the issue is happening. Put the code block below somewhere around [here](https://github.com/AMReX-Combustion/PeleLMeX/blob/a7364cb54a53a427abf9f522ea25bffda3ac782a/Source/PeleLMDiffusion.cpp#L1144) :
```
      std::string suffix = "_step"+std::to_string(m_nstep)+"_sdcIter"+std::to_string(m_sdcIter)+"_deltaTiter"+std::to_string(dTiter);
      WriteDebugPlotFile(GetVecOfConstPtrs(Tsave) , "Tsave" + suffix);
      WriteDebugPlotFile(GetVecOfConstPtrs(rhs) , "rhs" + suffix);
```
- Increase ` peleLM.deltaT_iterMax` to allow for better convergence, as the iterations appear to be converging, although slowly. Or set `peleLM.deltaT_crashIfFailing = 0` and hope that you can bypass the issue, but I wouldn't recommend continuing if the residuals don't improve after several timesteps.

**Reply by ZhaofanZhu** [reply]
Mr. Bruce, thanks for your guidance. I finally found the right way. I closed the reactions for a while, then opened them, and successfully avoided this error.

**Comment by drummerdoc** [comment]
I agree with Bruce's strategy.  Often times looking at the magnitude of the residuals over the domain can't be pretty revealing - is it getting hung up at the temperature jump or EB, or EB at the physical boundary corners?

It could be a bit convoluted though.  If the geometry details are preventing the creation of many multigrid V-cycle coarsening levels, the flat version may bypass that issue and allow many levels.  In the flat case, the "bottom solve" of the multigrid cycle would be nicely solved, but the wavy one may not be because with a much large bottom solve problem, it would take many more iterations than allowed for by default.  In this case, multigrid V-cycles would fail to converge to design spec.  Then the question is what are the settings you are using to tell the code what to do when the various solvers fail in this way?  Verbosity in the diffusion (and nodal and mac projections) solves might shed some light.

Perhaps another piece of information: what if you could set the wavelength of those ripples to be much longer, approaching the width of the domain?  Does the convergence of the deltaT iteration degrade smoothly with decreasing wavelength?  

I hope these suggestions help.

**Reply by ZhaofanZhu** [reply]
Thank you for your suggestions! I am still a beginner in this software. As I learn more about it, I'd love to explore it further.
